geomstats/geometry/complex_poincare_disk.py

```diff
@@ -0,0 +1,360 @@
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        """Compute the complex Poincaré disk inner-product.
+
+        Compute the inner-product of tangent_vec_a and tangent_vec_b
+        at point base_point using the Siegel Riemannian metric:
+
+        Parameters
+        ----------
+        tangent_vec_a : array-like, shape=[..., 1]
+            Tangent vector at base point.
+        tangent_vec_b : array-like, shape=[..., 1]
+            Tangent vector at base point.
+        base_point : array-like, shape=[..., 1]
+            Base point.
+
+        Returns
+        -------
+        inner_product : array-like, shape=[...,]
+            Inner-product.
+        """
+        inner_product_matrix = self.inner_product_matrix(base_point=base_point)
+        inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
+        inner_prod = gs.reshape(inner_prod, (-1,))
+        return inner_prod
+
+    def squared_norm(self, vector, base_point):
+        """Compute the square of the norm of a vector.
+
+        Squared norm of a vector associated to the inner product
+        at the tangent space at a base point.
+
+        Parameters
+        ----------
+        vector : array-like, shape=[..., 1]
+            Vector.
+        base_point : array-like, shape=[..., 1]
+            Base point.
+            Optional, default: None.
+
+        Returns
+        -------
+        sq_norm : array-like, shape=[...,]
+            Squared norm.
+        """
+        sq_norm = self.inner_product(vector, vector, base_point=base_point)
+        sq_norm = gs.real(sq_norm)
+        return sq_norm
+
+    @staticmethod
+    def exp(tangent_vec, base_point):
+        """Compute the complex Poincaré disk exponential map.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 1]
+            Tangent vector at base point.
+        base_point : array-like, shape=[..., 1]
+            Base point.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 1]
+            Riemannian exponential.
+        """
+        ndim = max(len(gs.shape(tangent_vec)), len(gs.shape(base_point)))
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=2)
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+        if base_point.shape[0] < tangent_vec.shape[0]:
+            base_point = (
+                gs.cast(gs.ones(tangent_vec.shape), dtype=base_point.dtype) @ base_point
+            )
+        elif base_point.shape[0] > tangent_vec.shape[0]:
+            tangent_vec = (
+                gs.cast(gs.ones(base_point.shape), dtype=tangent_vec.dtype)
+                @ tangent_vec
+            )
+        theta = gs.cast(gs.angle(tangent_vec), dtype=gs.get_default_cdtype())
+        s = 2 * gs.abs(tangent_vec) / (1 - gs.abs(base_point) ** 2)
+        num = gs.array(base_point)
+        exp_i_theta = gs.exp(1j * theta)
+        exp_minus_s = gs.cast(gs.exp(-s), dtype=gs.get_default_cdtype())
```

Comment:
unnecessary `cast`.

Reply:
If I do not cast, TensorFlow gives me the following error:

`tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]`